catalog/tabledesc: calculate KeysPerRow precisely for secondary indexes

yuzefovich opened the pull request:
Previously, we would say that a secondary index with a STORING clause
might have the same number of KVs per row as there are column families
in the table. However, this is not as precise as possible - namely, the
secondary index always uses the 0th column family, but 1st, 2nd, etc
column family is only used if at least one column from the family is in
the STORING clause (if not, that column family doesn't get a KV in the
secondary index). This observation allows us to compute `KeysPerRow`
precisely which then allows us to set smaller limits on the
BatchRequests in the fetchers.

Fixes: #72739

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72965)
<!-- Reviewable:end -->

yuzefovich committed:
catalog/tabledesc: calculate KeysPerRow precisely for secondary indexes

Previously, we would say that a secondary index with a STORING clause
might have the same number of KVs per row as there are column families
in the table. However, this is not as precise as possible - namely, the
secondary index always uses the 0th column family, but 1st, 2nd, etc
column family is only used if at least one column from the family is in
the STORING clause (if not, that column family doesn't get a KV in the
secondary index). This observation allows us to compute `KeysPerRow`
precisely which then allows us to set smaller limits on the
BatchRequests in the fetchers.

Release note: None

yuzefovich commented:
cc @nvanbenschoten

ajwerner commented:
These changes scare me given what we now know about bugs in index encodings. I definitely defer to @cockroachdb/sql-queries folks when it comes to row encoding sorts of things.

yuzefovich commented:
At least currently `KeysPerRow` is only used for a single purpose - to calculate the key limit for the first BatchRequest issued by the SQL fetchers, and if we get this wrong, it'll only have a performance impact (like we'll need to send an extra BatchRequest because not all column families for all needed rows have been filled), so I'm not that concerned about the correctness here.

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72965#-:-MpCTgfr56x9W54q1uzt:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
